### DISCUSSION TITLE
Backup preconditions (e.g. machine running on wall power - not battery)

### DISCUSSION THREAD

**Discussion by andrew-hardin** [borgbase/vorta#1153] [Ideas]
I run periodic Borg backups using a few hand rolled scripts that are executed via `systemd`. Vorta is almost feature complete for my use case. The only missing feature is the concept of preconditions -- a check that must pass prior to running a backup, for example:

- Is the machine plugged into the wall, or, is it running on battery power? Only run backups on wall power.

Is there any interest in adding a power precondition to Vorta?

**Comment by m3nu** [comment]
Yes, that would be a nice addition for sure. Have you already tried using a pre-backup command? If it doesn't exit with status `0` then the backup wouldn't run.

We also have preconditions for networks. Under *Schedule > Networks* "Don't run on metered networks".

For a power precondition, I'd add it under *Schedule* in the right tab. Implementation wise, we'd need at least Linux and macOS. Usually we do an abstract class and then subclasses for each OS. See the `network_status` package for an example.

**Reply by andrew-hardin** [reply]
I haven't tried the pre-backup command, but that looks like it would work -- that's a powerful feature to have arbitrary pre/post backup hooks.

I've taken a quick pass at a notional power tab. The default behavior would be to ignore any power status (i.e. maintain the status quo).
![image](https://user-images.githubusercontent.com/16496326/148164031-177fcc42-c1e8-48b9-8b13-a93a134e94d2.png)

For the implementation, `psutil.sensors_battery()` can yield the presence of a battery and estimated remaining time. I've verified the method works on Ubuntu, but can't verify on macOS. Perhaps outsourcing that logic to `psutil` alleviates the need for an interface with per-operating system implementations?
